// better-youtube-api/out/entities/channel.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A YouTube channel.
 */
class Channel {
    constructor(youtube, data) {
        this.youtube = youtube;
        this.data = data;
        this._init(data);
    }
    _init(data) {
        if (data.kind === 'youtube#channel' && data.status.isLinked) {
            const channel = data;
            this.id = channel.id;
            this.country = channel.snippet.country;
            this.language = channel.snippet.defaultLanguage;
            this.views = Number(channel.statistics.viewCount);
            this.commentCount = Number(channel.statistics.commentCount);
            this.banners = channel.brandingSettings.image;
            // Unknown behavior
            /* istanbul ignore next */
            if (channel.brandingSettings.channel) {
                this.keywords = channel.brandingSettings.channel.keywords ? channel.brandingSettings.channel.keywords.split(' ') : [];
                this.featuredChannels = channel.brandingSettings.channel.featuredChannelsUrls ?
                    channel.brandingSettings.channel.featuredChannelsUrls.map(id => `https://www.youtube.com/channel/${id}`) : [];
            }
            if (!channel.statistics.hiddenSubscriberCount) {
                this.subCount = Number(channel.statistics.subscriberCount);
            }
            else {
                this.subCount = -1;
            }
        }
        else if (data.kind === 'youtube#searchResult') {
            this.id = data.id.channelId;
            // Impossible to test
            /* istanbul ignore next */
            this.liveStatus = data.snippet.liveBroadcastContent !== 'none' ? data.snippet.liveBroadcastContent : false;
        }
        else {
            throw new Error(`Invalid channel type: ${data.kind}`);
        }
        this.url = `https://youtube.com/channel/${this.id}`;
        this.profilePictures = data.snippet.thumbnails;
        this.dateCreated = new Date(data.snippet.publishedAt);
        this.name = data.snippet.title;
        this.about = data.snippet.description;
        this.full = data.kind === 'youtube#channel';
    }
    /**
     * Posts a comment on the channel's discussion tab.
     * Must be using an access token with correct scopes.
     * @param text The text of the comment.
     */
    /* istanbul ignore next */
    postComment(text) {
        return __awaiter(this, void 0, void 0, function* () {
            const comment = yield this.youtube.oauth.postComment(text, this.id);
            if (this.comments !== undefined) {
                this.comments.push(comment);
            }
            else {
                this.comments = [comment];
            }
            return comment;
        });
    }
    /**
     * Fetches this channel and reassigns this object to the new channel object.
     * Only useful if `this.full` is false, or if you want updated channel info.
     */
    fetch() {
        return __awaiter(this, void 0, void 0, function* () {
            const channel = yield this.youtube.getChannel(this.id);
            return Object.assign(this, channel);
        });
    }
    /**
     * Fetches the channel's videos and assigns them to the `Channel#videos` property.
     */
    fetchVideos() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!(this.data.contentDetails)) {
                yield this.fetch();
            }
            const videos = yield this.youtube.getPlaylist(this.data.contentDetails.relatedPlaylists.uploads);
            this.videos = videos;
            return this.videos;
        });
    }
    /**
     * Fetches the channel's discussion tab comments and assigns them to Channel#comments.
     * @param maxResults The maximum amount of comments to fetch
     */
    fetchComments(maxResults = -1) {
        return __awaiter(this, void 0, void 0, function* () {
            this.comments = yield this.youtube.getChannelComments(this.id, maxResults);
            return this.comments;
        });
    }
    /**
     * Fetches the channel's playlists and assigns them to Channel#playlists.
     * @param maxResults The maximum amount of playlists to fetch
     */
    fetchPlaylists(maxResults = -1) {
        return __awaiter(this, void 0, void 0, function* () {
            this.playlists = yield this.youtube.getChannelPlaylists(this.id, maxResults);
            return this.playlists;
        });
    }
}
exports.Channel = Channel;
/**
 * The name of the endpoint used for this entity.
 */
Channel.endpoint = 'channels';
/**
 * The parts to request for this entity.
 */
Channel.part = 'snippet,contentDetails,statistics,status,brandingSettings';
/**
 * The fields to request for this entity.
 */
Channel.fields = 'items(kind,id,contentDetails(relatedPlaylists(uploads)),statistics(subscriberCount,commentCount,viewCount,hiddenSubscriberCount),' +
    'snippet(title,description,thumbnails,publishedAt,country,defaultLanguage),brandingSettings(image,channel(keywords,featuredChannelsUrls)),status(isLinked))';
